http_parser.c

```diff
@@ -472,6 +483,92 @@ static struct {
 
 int http_message_needs_eof(const http_parser *parser);
 
+const char* find_crlf(const char* p, const char* data, size_t len);
+
+
+#if defined(__SSE2__) && defined(__GNUC__) 
+
+#include <emmintrin.h>
+
+const char* find_crlf(const char* p, const char* data, size_t len) {
+  const char* lastp = MIN(data + len, HTTP_MAX_HEADER_SIZE + p);
+
+  int32_t result = 0;
+  __m128i v1, v2;
+  /* [ c, 0, 0, 0, 0, 0 .. 0 ] */
+  __m128i vCR = _mm_set_epi32(0x0a0a0a0a, 0x0a0a0a0a,0x0a0a0a0a, 0x0a0a0a0a);
+
+  /* [ c, 0, 0, 0, 0, 0 .. 0 ] */
+  __m128i vLF = _mm_set_epi32(0x0d0d0d0d, 0x0d0d0d0d,0x0d0d0d0d, 0x0d0d0d0d);
+
+  size_t alignment = (long) p & 15;
+  p -= alignment;
+
+  v1 = *(__m128i*)p;
+  v2 = _mm_cmpeq_epi8(vCR, v1);
+  v1 = _mm_cmpeq_epi8(vLF, v1);
+  v2 = _mm_or_si128(v1, v2);
+  result = _mm_movemask_epi8(v2);
+
+  v1 = *(__m128i*)(p + 16);
+  v2 = _mm_cmpeq_epi8(vCR, v1);
+  v1 = _mm_cmpeq_epi8(vLF, v1);
+  v2 = _mm_or_si128(v1, v2);
+  result = _mm_movemask_epi8(v2) << 16 | result;
+  result = (result >> alignment) << alignment;
+
+  if (!result) {
+    while(!result && lastp >= p+32) {
+      p += 32;
+      v1 = *(__m128i*)p;
+      v2 = _mm_cmpeq_epi8(vCR, v1);
+      v1 = _mm_cmpeq_epi8(vLF, v1);
+      v2 = _mm_or_si128(v1, v2);
+      result = _mm_movemask_epi8(v2);
+
+      v1 = *(__m128i*)(p+16);
+      v2 = _mm_cmpeq_epi8(vCR, v1);
+      v1 = _mm_cmpeq_epi8(vLF, v1);
+      v2 = _mm_or_si128(v1, v2);
+      result = _mm_movemask_epi8(v2) << 16 | result;
```

Comment:
These four lines are repeated four times; you could add a little `inline static int get_crlf_mask(const char*)` method maybe.

Reply:
Thanks for the comments! i made most of the changes and tested, they are all working, as for the inline function, I am not quite sure if that works on every platform's compiler's c dialect though

Reply:
Awesome! Thanks. (Did you push the changes up?)

As for `inline static`, I saw the line below saying it's a C99 project, which means it's in the standard. GCC has supported it since 4.2. Obviously unnecessary, just nice cleanup.
https://github.com/nodejs/http-parser/blob/master/http_parser.gyp#L65

Reply:
nope, I have yet to push the change, will be doing that tonight( asia time, working hours :P )

Oh what I kinda worry about is that whether the other compilable platforms has the dialect turned on, cause this is my first contribution I am not quite sure, if you are sure about it, let me make that change too!

Reply:
Or perhaps anybody else who might be able to shed a light on whether this is safe to use, or should I write a macro instead?

Reply:
Why don't make a regular function? Will it be a big overhead?

Reply:
> whether the other compilable platforms has the dialect turned on

GCC and Clang default to gnu11, which supports `inline`. MSVC is set to compile this package as C++. Since you're testing for `__GNUC__`, I think that means it would be safe. But... in any case, I just tested GCC 4.4.7, 5.1, 6.1 and 7.2, and LLVM Clang 4.0.0 and 5.0.0, and they all automatically inline the static helper function without the `inline` keyword in this case.

Reply:
Yup, I just did some test and compiled them last night too, most of them seem to inline it automatically anyways.